### ISSUE TITLE
ZIO 2.0: Smart constructors

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#5241]
Every data type in ZIO (ZIO, ZManaged, ZStream, etc.) has a variety of constructor functions that are designed to "up convert" some weaker type into the target type. Typically, these converter functions are named `fromXYZ`, e.g. `ZIO.fromEither`, `ZStream.fromZIO`, etc.

While these are precise, it could be useful to specially designate the `XYZ.apply` method of the companion object as a "smart constructor", which will intelligently choose the most likely constructor based on the input type.

This would allow you to write, for example:

```scala
ZStream(List(1, 2, 3))
ZStream(zioEffect)
ZStream(queue)
...
ZIO(Right(23))
ZIO(Try(123))
ZIO(future)
...
```

I prototyped this feature but have not extensively tested it.

```scala
import scala.util.Try 
sealed trait ZIO[-R, +E, +A]
object ZIO {
  def apply[Input](input: => Input)(implicit c: ZIOConstructor[Input]): c.Out = c.make(input)
  
  def fromTry[A](t: Try[A]): ZIO[Any, Throwable, A] = ???
  def fromEither[E, A](e: Either[E, A]): ZIO[Any, E, A] = ???
  def attempt[A](action: => A): ZIO[Any, Throwable, A] = ???
  def fromFunction[A, B](f: A => B): ZIO[A, Nothing, B] = ???
}
sealed trait ZIOConstructor[-Input] {
  type Out
  
  def make(input: => Input): Out
}
object ZIOConstructor extends ZIOConstructorLowPriority {
//   implicit def FunctionConstructor[A, B]: WithOut[A => B, ZIO[A, Nothing, B]] = new ZIOConstructor[A => B] {
//     type Out = ZIO[A, Nothing, B]
    
//     def make(input: => (A => B)): ZIO[A, Nothing, B] = ZIO.fromFunction(input)
//   }  
  implicit def EitherConstructor[E, A]: WithOut[Either[E, A], ZIO[Any, E, A]] = new ZIOConstructor[Either[E, A]] {
    type Out = ZIO[Any, E, A]
    
    def make(input: => Either[E, A]): ZIO[Any, E, A] = ZIO.fromEither(input)
  }  
  implicit def TryConstructor[A]: WithOut[Try[A], ZIO[Any, Throwable, A]] = new ZIOConstructor[Try[A]] {
    type Out = ZIO[Any, Throwable, A]
    
    def make(input: => Try[A]): ZIO[Any, Throwable, A] = ZIO.fromTry(input)
  }
}
trait ZIOConstructorLowPriority {
  type WithOut[-In, Out0] = ZIOConstructor[In] { type Out = Out0 }
  
  implicit def AttemptConstructor[A]: WithOut[A, ZIO[Any, Throwable, A]] = new ZIOConstructor[A] {
    type Out = ZIO[Any, Throwable, A]
    
    def make(input: => A): ZIO[Any, Throwable, A] = ZIO.attempt(input)
  }
}

object example {
  val zio1 : ZIO[Any, Int, String] = ZIO(Left(12))
  val zio2 : ZIO[Any, Throwable, Int] = ZIO(Try(12))
  val zio3 : ZIO[Any, Throwable, Boolean] = ZIO(true)
}
```

**Comment by adamgfraser** [comment]
I will work on this.

**Comment by jdegoes** [comment]
@adamgfraser Awesome. Could be a big win especially for new users, as well as improve readability of dense "interop" code.

I couldn't make `Function` work, maybe something stupid (or possibly related to bugs around the contravariance of the `A` in Function).

**Comment by adamgfraser** [comment]
Yes the other way we could do this is to provide `toZYZ` operators on everything. That might infer better in some cases.

**Comment by jdegoes** [comment]
Unfortunately, you need implicits and therefore implicit imports (also the syntax is not as nice).

**Comment by adamgfraser** [comment]
Agreed.

**Comment by adamgfraser** [comment]
Yes I think the issue with `Function` is caused by the contravariance on `Input`. If I make that `Invariant` the example works. That creates a little bit of an issue because it should naturally be contravariant and not having the variance requires us to define separate instances for types like `Left` and `Right` but I think that is potentially solvable without visibility to the user.

**Comment by jdegoes** [comment]
Yes, since the number of data types is small and finite, it's probably OK for us to "emulate" what the contravariance buys us.

**Comment by adamgfraser** [comment]
Did you look at overloading plus extension methods for the lower priority constructors? Seems like that would be the other way to do this.

**Comment by jdegoes** [comment]
I don't think overloading and implicits will work together (I've tried that
before).

**Comment by joroKr21** [comment]
I like it. You can make `ZIOConstructor` invariant if you use it like this:
```scala
def apply[Input](input: => Input)(implicit c: ZIOConstructor[_ >: Input]): c.Out = c.make(input)
```

But you still have to provide the function parameter type explicitly.
And I think @adamgfraser meant using overloading without implicits:

```scala
object ZIO {
  def apply[A](t: Try[A]): ZIO[Any, Throwable, A] = ???
  def apply[E, A](e: Either[E, A]): ZIO[Any, E, A] = ???
  def apply[A](action: => A): ZIO[Any, Throwable, A] = ???
  def apply[A, B](f: A => B): ZIO[A, Nothing, B] = ???
}

object example {
  val zio1: ZIO[Any, Int, String] = ZIO(Left(12))
  val zio2: ZIO[Any, Throwable, Int] = ZIO(Try(12))
  val zio3: ZIO[Any, Throwable, Boolean] = ZIO(true)
  val zio4: ZIO[Int, Nothing, String] = ZIO(_.toString)
}
```

It should work fairly well from Scala 2.13 onwards because overloading was improved to support the new collections design and even infer function parameter types.

**Comment by adamgfraser** [comment]
Closed by #5246.
